src/dotnet/ToolPackage/ToolPackageObtainer.cs

```diff
@@ -31,9 +32,54 @@ IProjectRestorer projectRestorer
             _projectRestorer = projectRestorer ?? throw new ArgumentNullException(nameof(projectRestorer));
             _toolsPath = toolsPath;
             _offlineFeedPath = offlineFeedPath;
+
+            _stageDirectory = _toolsPath.WithSubDirectories(".stage", Path.GetRandomFileName());
```

Comment:
Need to move down to CreateObtainTransaction method, make it a local, and let closure capture it